doc/papers/joss/paper.md

```diff
@@ -0,0 +1,115 @@
+(or their gradients) on the nodes, and $D$ is the user-defined pointwise function
+describing the physics of the problem at the quadrature points, also called the
+QFunction. QFunctions, which can either be defined by the user or selected from a
+gallery of available built-in functions in the library, are pointwise functions
+that do not depend on element resolution, topology, or basis degree (selectable
+at run time). This easily allows $hp$-refinement studies (where $h$ commonly denotes the average element size and $p$ the polynomial degree of the basis functions in 1D) and $p$-multigrid solvers. libCEED also supports composition of different operators for multiphysics problems and mixed-element meshes (see Fig. \ref{fig:schematic}).
+
+![A schematic of element restriction and basis applicator operators for
+elements with different topology. This sketch shows the independence of QFunctions
+(in this case representing a Laplacian) element resolution, topology, or basis degree.\label{fig:schematic}](img/QFunctionSketch.pdf)
+
+To achieve high performance, libCEED can take advantage of a tensor-product
+finite-element basis and quadrature rule to apply the action of the basis
+operator $B$ or, alternatively, efficiently operate on bases that are defined
+on arbitrary-topology elements. Furthermore, the algebraic decomposition described in
+Eq. (\ref{eq:decomposition}) can represent either linear/nonlinear or
+symmetric/asymmetric operators and exposes opportunities for device-specific
+optimizations.
+
+![libCEED is a low-level API for finite element codes, that has specialized implementations
+(backends) for heterogeneous architectures.\label{fig:libCEEDBackends}](img/libCEEDBackends.pdf)
+
+Fig. \ref{fig:libCEEDBackends} shows a subset of the backend implementations available
+in ``libCEED``. GPU implementations are available via pure CUDA as well as the OCCA
+and MAGMA libraries. CPU implementations are available via pure C and AVX intrinsics
+as well as the LIBXSMM library. libCEED provides a unified interface, so that users
+only need to write a single source code and can select the desired specialized
+implementation at run time. Moreover, each process or thread can instantiate an
+arbitrary number of backends.
+
+# Performance Benchmarks
+
+The Center for Efficient Exascale Discretizations (CEED), part of the Exascale Computing Project (ECP) uses Benchmark Problems (BPs) to test and compare the performance of high-order finite element implementations. In Fig. \ref{fig:Skylakeselfxsmm}, we analyze the performance of libCEED's LIBXSMM backend on a Poisson's problem. We measure performance over 20 iterations of unpreconditioned Conjugate Gradient (CG) and plot throughput, versus execution time per iteration, for different values of the polynomial degree $p$.
+
+![\texttt{xsmm/serial} backend on a Skylake
+(2x Intel Xeon Platinum 8180M CPU 2.50GHz).\label{fig:Skylakeselfxsmm}](img/plot_libCEED_PETScBP3_cpuselfxsmmserial_N001_pn56_time.pdf)
```

Comment:
I think the /xsmm/* and /cuda/gen bechnmarks are good here. BP3?

Reply:
It was just an old figure from another 2-page summary produced for a conference submission. Totally up to us to decide which and how many results to show (which was one of the checkboxes above)

Reply:
Cool. I'm just typing out thoughts as they occur to me, but you were totally ahead of me on this one.

Reply:
I think that we have enough small changes in performance that we should do fresh runs. I'm thinking BP3 or BP1 + BP3 in the following configurations:

1) `/gpu/cuda/gen` + V100
2) `/cpu/self/xsmm/serial` + Skylake Platinum
3) `/cpu/self/xsmm/blocked` + Skylake Platinum

The benchmark script uses the `bpsraw` right now, but I'm honestly not sure if `bpsraw` or `bps` runs faster.

Reply:
By a very quick check on my machine, without even running the whole battery for all degrees like in the benchmark script, but only one configuration, the `bpsraw` is actually slower than the `bps`

Reply:
I think I used `bpsraw` at first because `bps` originally had a pretty big setup time for the mesh. Since its faster, I think we can switch over (though, we'd have to update the script to run based on number of box faces or compute faces out of local dofs desired.

Reply:
I can't upload the log files for some reason now in this thread.. But yes, for my very quick test on my machine, from PETSc' l`-og_view` summary (the Max column) I've got:
bpsraw in serial:
Time (sec):           5.403e-02
bps in serial:
Time (sec):           2.709e-02

bpsraw in parallel (with `mpiexec -n 2`):
Time (sec):           3.220e-02
bps in parallel (with `mpiexec -n 2`):
Time (sec):           1.685e-02

Reply:
They were just run as `./bpsraw -problem bp3` and `./bps -problem bp3`, and ` mpiexec -n 2 ./bpsraw -problem bp3` and ` mpiexec -n 2 ./bps -problem bp3`, respectively, without any other specified option

Reply:
I still think that `bps` is faster, but the default sizes don't match.
Closer lvec sized test: (fiddling with `-local` to get `-cells` is a pain)
```
[jeremy@sinensis petsc]$ ./bpsraw -problem bp3 -degree 2 -local 9261

-- CEED Benchmark Problem 3 -- libCEED + PETSc --
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/xsmm/blocked
    libCEED Backend MemType            : host
    libCEED User Requested MemType     : none
  Mesh:
    Number of 1D Basis Nodes (p)       : 3
    Number of 1D Quadrature Points (q) : 4
    Global nodes                       : 10773
    Process Decomposition              : 1 1 1
    Local Elements                     : 1170 = 9 10 13
    Owned nodes                        : 10773 = 19 21 27
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 9
    Final rnorm                        : 2.149739e-12
  Performance:
    Pointwise Error (max)              : 2.327021e-03
    CG Solve Time                      : 0.00627679 (0.00627679) sec
[jeremy@sinensis petsc]$ ./bps -problem bp3 -cells 10,10,10 -degree 2

-- CEED Benchmark Problem 3 -- libCEED + PETSc --
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/xsmm/blocked
    libCEED Backend MemType            : host
    libCEED User Requested MemType     : none
  Mesh:
    Number of 1D Basis Nodes (p)       : 3
    Number of 1D Quadrature Points (q) : 4
    Global nodes                       : 6859
    Owned nodes                        : 6859
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 8
    Final rnorm                        : 1.379813e-12
  Performance:
    Pointwise Error (max)              : 5.054563e-03
    CG Solve Time                      : 0.00467284 (0.00467284) sec
```

Reply:
Besides the solve time... It's also interesting the double-sized error, isn't it? 🤔

Reply:
Maybe because of the extra CG iteration the bpsraw error goes down slightly. As we increase mesh size I'd imagine the difference goes down.

Reply:
Yeah maybe... Up to us whether we want to modify the benchmark script to run `bps` instead of `bpsraw` and compare results for all problem sizes to make sure that `bps` actually perform better

Reply:
It makes sense for `bps` (after initialization) to be slightly faster because of how it handles boundary conditions.

Reply:
> I think that we have enough small changes in performance that we should do fresh runs. I'm thinking BP3 or BP1 + BP3 in the following configurations:
> 
> 1. `/gpu/cuda/gen` + V100
> 2. `/cpu/self/xsmm/serial` + Skylake Platinum
> 3. `/cpu/self/xsmm/blocked` + Skylake Platinum
> 
Hey, @jeremylt @jedbrown I have run the benchmarks using the `petsc-bps` on CU Summit's skylake partition (the only Skylake node I have access to). The [results](https://drive.google.com/open?id=1GCgFZG4_Oi6Ad3eF6dHBJ11edwGbBh5_) are not better than what we had already that @jedbrown tested last summer (I don't know on which machine).

Please let me know what other tests I could run, as I was hoping we could select some of the updated results also for the SciPy submission, which is coming up. Thanks.

Reply:
I think CU has Skylake Golds, so they won't be as good as @jedbrown's runs on Skylake Platinums.

Reply:
Yeah, their documentation on hardware is spotty.

Here is what I've found:
https://www.colorado.edu/rc/news/past/summitcondo

Their Summit specs are out of date:
https://www.colorado.edu/rc/resources/summit/specifications

Reply:
You can always use `cat /proc/cpuinfo`. 😁